B(1) = +½ set 3: harmonic numbers at complex arguments

Parcly-Taxel opened the pull request:
#### References to other Issues or PRs

Part 3 of the changes formerly included in #23926, building upon #23973. Fixes #23977.

#### Brief description of what is fixed or changed

Unlike #23984 (which this PR is independent of) this function generalisation does not appear in Luschny, but it does appear in several places, including Wikipedia:
$$H_{n,m}=\begin{cases}\psi(n+1)+\gamma&m=1\\\\ \zeta(m)-\zeta(m,n+1)&m\ne1\end{cases}$$

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* functions
    * `harmonic()` now accepts complex arguments.
<!-- END RELEASE NOTES -->

Parcly-Taxel committed:
functions: Extend harmonic numbers to complex arguments

There is a generally accepted extension of the harmonic numbers to
complex arguments based on zeta functions:

H_{n, m} = ζ(m) - ζ(m, n + 1)

Of course this does not apply if m = 1, but in that case we can take the
limiting expression

H_n = ψ(n + 1) + γ

It also does not apply if n is a negative integer – the harmonic number
remains undefined there.

In addition, the existing rewrite from harmonic() to polygamma() is
actually only valid if m is a positive integer; a corresponding guard
has been added.

asmeurer commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -950,9 +957,16 @@ def _eval_rewrite_as_tractable(self, n, m=1, limitvar=None, **kwargs):
+            if m == 1:
+                res = mp.harmonic(n)
+            else:
+                res = mp.zeta(m) - mp.zeta(m, n+1)
```
Why not use recursive evalf on sympy.zeta?

Parcly-Taxel replied:
Let's say `harmonic().evalf()` looks like this:
```python
    def _eval_evalf(self, prec):
        from sympy.functions.special.zeta_functions import zeta
        if all(i.is_number for i in self.args):
            return self.rewrite(zeta)._eval_evalf(prec)
```

Then one of the test cases I added will fail:
https://github.com/sympy/sympy/blob/96a51c83529162fec3171b451a15764bafba80b3/sympy/functions/combinatorial/tests/test_comb_numbers.py#L296

The rewritten Piecewise expression will not resolve to the special case despite `m` being exactly one and `zeta()` will be numerically evaluated at first argument 1, resulting in a pole error.

```
Traceback (most recent call last):
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/functions/combinatorial/tests/test_comb_numbers.py", line 296, in test_harmonic_evalf
    assert str(harmonic(7.0, 1.0).evalf(n=10)) == '2.592857143'
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/function.py", line 475, in __new__
    result = result.evalf(prec_to_dps(pr))
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1659, in evalf
    result = evalf(self, prec + 4, options)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1501, in evalf
    xe = x._eval_evalf(prec)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/functions/combinatorial/numbers.py", line 962, in _eval_evalf
    return self.rewrite(zeta)._eval_evalf(prec)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/basic.py", line 1943, in rewrite
    return self._rewrite(pattern, rule, method, **hints)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/basic.py", line 1955, in _rewrite
    rewritten = meth(*args, **hints)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/functions/combinatorial/numbers.py", line 917, in _eval_rewrite_as_zeta
    (zeta(m) - zeta(m, n+1), True))
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/cache.py", line 72, in wrapper
    retval = cfunc(*args, **kwargs)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/function.py", line 475, in __new__
    result = result.evalf(prec_to_dps(pr))
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1659, in evalf
    result = evalf(self, prec + 4, options)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/evalf.py", line 1501, in evalf
    xe = x._eval_evalf(prec)
  File "/home/parclytaxel/Documents/Computing/Repos/sympy/sympy/core/function.py", line 593, in _eval_evalf
    v = func(*args)
  File "/usr/local/lib/python3.10/dist-packages/mpmath/functions/zeta.py", line 533, in zeta
    return ctx._zeta(s, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/mpmath/ctx_mp_python.py", line 998, in f
    return ctx.make_mpf(mpf_f(x._mpf_, prec, rounding))
  File "/usr/local/lib/python3.10/dist-packages/mpmath/libmp/gammazeta.py", line 984, in mpf_zeta
    return mpf_zeta_int(to_int(s), prec, rnd)
  File "/usr/local/lib/python3.10/dist-packages/mpmath/libmp/gammazeta.py", line 917, in mpf_zeta_int
    raise ValueError("zeta(1) pole")
ValueError: zeta(1) pole
```

Parcly-Taxel replied:
@asmeurer is there a way to use the recursive evalf _and_ have Piecewise recognise the special case $m=1.0$ properly?

oscarbenjamin commented on sympy/functions/combinatorial/tests/test_comb_numbers.py:
```diff
@@ -292,18 +292,23 @@ def test_harmonic_rational():
+    assert str(harmonic(7.0, 1.0).evalf(n=10)) == '2.592857143'
+    assert str(harmonic(1, pi).evalf(n=10)) == '1.000000000'
+    assert str(harmonic(2, pi).evalf(n=10)) == '1.113314732'
+    assert str(harmonic(1000.0, pi).evalf(n=10)) == '1.176241563'
```
I guess there should be some tests for complex numbers here.

oscarbenjamin replied:
Okay, plotting works now after the last change so I can see a bit better how the function looks.

oscarbenjamin commented:
Perhaps this should be handled in evalf:
```python
In [11]: harmonic(-1, 2).evalf()
...
ZeroDivisionError 
```

oscarbenjamin commented:
What exactly is going on here:
```python
In [35]: harmonic(-S(1)-1e-10, I).evalf()
Out[35]: 11.8198472989154 + 19.8942922807166⋅ⅈ

In [36]: harmonic(-S(1)+1e-10, I).evalf()
Out[36]: 0.51078191516229 + 0.859710320595213⋅ⅈ

In [37]: harmonic(-S(1), I).evalf()
Out[37]: -1.00000000000000
```

oscarbenjamin commented:
And this:
```python
In [45]: harmonic(-S(1), I-1e-10).evalf()
Out[45]: nan

In [46]: harmonic(-S(1), I+1e-10).evalf()
Out[46]: -1.00000000000000
```

oscarbenjamin commented:
Also
```python
In [9]: harmonic(-1, 1).evalf()

ValueError: polygamma pole
> /38venv/lib/python3.8/site-packages/mpmath/libmp/gammazeta.py(667)mpf_psi0()
    665         if x == fninf or x == fnan: return fnan
    666     if x == fzero or (exp >= 0 and sign):
--> 667         raise ValueError("polygamma pole")
    668     # Near 0 -- fixed-point arithmetic becomes bad
    669     if exp+bc < -5:
```

Parcly-Taxel committed:
functions: Add harmonic NaN check and complex tests

oscarbenjamin commented:
Is it possible to make `lambdify` work for `harmonic`?
```
In [47]: lambdify(x, harmonic(x))(1)
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-47-ead5f6efcb7f> in <module>
----> 1 lambdify(x, harmonic(x))(1)

<lambdifygenerated-4> in _lambdifygenerated(x)
      1 def _lambdifygenerated(x):
----> 2     return harmonic(x)

NameError: name 'harmonic' is not defined
```

Parcly-Taxel committed:
Merge branch 'master' into harmonic

Parcly-Taxel committed:
Add lambdify support for harmonic()

Parcly-Taxel commented:
> Is it possible to make `lambdify` work for `harmonic`?

Yes.

oscarbenjamin commented:
Is it possible to define things like derivatives for `harmonic`? This rewrite indicates that it might be in some cases:
```python
In [44]: harmonic(x, y).diff(x)
Out[44]: 
∂                 
──(harmonic(x, y))
∂x                

In [45]: harmonic(x, y).rewrite(zeta).diff(x)
Out[45]: 
⎧polygamma(1, x + 1)  for y = 1
⎨                              
⎩ y⋅ζ(y + 1, x + 1)   otherwise
```
I don't know a lot about how the series code works but maybe something like this can be used for series:
```python
In [47]: harmonic(x, 1).rewrite(zeta).series(x)
Out[47]: 
 2      2                    4  3    4                    6  5        
π ⋅x   x ⋅polygamma(2, 1)   π ⋅x    x ⋅polygamma(4, 1)   π ⋅x     ⎛ 6⎞
──── + ────────────────── + ───── + ────────────────── + ───── + O⎝x ⎠
 6             2              90            24            945         

In [48]: harmonic(x, 1).series(x)
Out[48]: 
                             ⎛  2             ⎞│         ⎛  3             ⎞│         ⎛  4          
                           2 ⎜ d              ⎟│       3 ⎜ d              ⎟│       4 ⎜ d           
                          x ⋅⎜───(harmonic(ξ))⎟│      x ⋅⎜───(harmonic(ξ))⎟│      x ⋅⎜───(harmonic(
                             ⎜  2             ⎟│         ⎜  3             ⎟│         ⎜  4          
  ⎛d              ⎞│         ⎝dξ              ⎠│ξ=0      ⎝dξ              ⎠│ξ=0      ⎝dξ           
x⋅⎜──(harmonic(ξ))⎟│    + ───────────────────────── + ───────────────────────── + ─────────────────
  ⎝dξ             ⎠│ξ=0               2                           6                           24   

   ⎞│         ⎛  5             ⎞│           
   ⎟│       5 ⎜ d              ⎟│           
ξ))⎟│      x ⋅⎜───(harmonic(ξ))⎟│           
   ⎟│         ⎜  5             ⎟│           
   ⎠│ξ=0      ⎝dξ              ⎠│ξ=0    ⎛ 6⎞
──────── + ───────────────────────── + O⎝x ⎠
                      120              
```
This limit gives a RecursionError in this PR:
```python
In [54]: harmonic(x, y).limit(x, 0)
---------------------------------------------------------------------------
RecursionError

~/current/sympy/sympy.git/sympy/core/basic.py in rewrite(self, deep, *args, **hints)
   1941         # hereafter, empty pattern is interpreted as all pattern.
   1942 
-> 1943         return self._rewrite(pattern, rule, method, **hints)
   1944 
   1945     def _rewrite(self, pattern, rule, method, **hints):

~/current/sympy/sympy.git/sympy/core/basic.py in _rewrite(self, pattern, rule, method, **hints)
   1953             meth = getattr(self, method, None)
   1954             if meth is not None:
-> 1955                 rewritten = meth(*args, **hints)
   1956             else:
   1957                 rewritten = self._eval_rewrite(rule, args, **hints)

~/current/sympy/sympy.git/sympy/functions/combinatorial/numbers.py in _eval_rewrite_as_tractable(self, n, m, limitvar, **kwargs)
    990     def _eval_rewrite_as_tractable(self, n, m=1, limitvar=None, **kwargs):
    991         from sympy.functions.special.gamma_functions import polygamma
--> 992         return self.rewrite(polygamma).rewrite("tractable", deep=True)
    993 
    994     def _eval_evalf(self, prec):

... last 3 frames repeated, from the frame below ...

~/current/sympy/sympy.git/sympy/core/basic.py in rewrite(self, deep, *args, **hints)
   1941         # hereafter, empty pattern is interpreted as all pattern.
   1942 
-> 1943         return self._rewrite(pattern, rule, method, **hints)
   1944 
   1945     def _rewrite(self, pattern, rule, method, **hints):

RecursionError: maximum recursion depth exceeded while calling a Python object
```

oscarbenjamin commented:
I don't know if all of these functions should be defining things like `_eval_is_meromorphic` or `_singularities`.

Parcly-Taxel committed:
Define basic calculus on the harmonic numbers

Parcly-Taxel commented:
> I don't know if all of these functions should be defining things like `_eval_is_meromorphic` or `_singularities`.

All the little problems you pointed out in the previous post have been fixed. But we shouldn't be bikeshedding over function features here – my new polygamma code depends on this PR being merged. Just merge this, and ask for features if you _really_ need them.

Parcly-Taxel committed:
It no longer XFAILs

Parcly-Taxel committed:
One import back

Parcly-Taxel committed:
…and one import forwards

oscarbenjamin commented:
> Just merge this, and ask for features if you _really_ need them.

Ultimately all of these things are needed. The work that you propose to "implement" these functions in the end requires everything from differentiation, to series to simplification and so on. Adding numerical evaluation and rewrites is an important first step but it still leaves a lot of work to be picked up in future. If you don't intend to do that work yourself then it will mean that someone else will need to do it in future.

When I raise comments as above it doesn't mean that all of these things are needed before the PR can be merged. Rather I am asking questions (because I want to know the answer) or pointing to things that you might not have considered. The discussion itself can be as valuable as the changes to the code when someone in the future looks back at this (especially if some of the work is left unfinished).

Note that at least one of the points above (about limits) was in some sense a regression compared to master. What we do need to do before merging any PR is try to ensure that we aren't introducing any new bugs.

Parcly-Taxel committed:
functions: Tiered tractable rewriting for harmonic()

We first try to rewrite in terms of polygamma(), which is only possible
if the second argument m is a positive integer. If the rewrite fails we
fall back to rewriting in terms of zeta().

Parcly-Taxel commented:
> When I raise comments as above it doesn't mean that all of these things are needed before the PR can be merged. Rather I am asking questions because I want to know the answer or pointing to things that you might not have considered.

Right, all fixed. What else do I need before merging?

You could have realised I was getting tired of your comments when I fumbled modifying the imports after I changed the test cases. I've been coding well past midnight for several days now while attending full-day training courses at my company.

While I am going to try and accommodate everything you heap on me, please remember that I am not an all-powerful genie ([despite my ponysona being a genie](https://derpibooru.org/images/2906946)).

asmeurer commented:
There's no rush to get these changes in. If you feel you are overworking to get this done then you should scale back and not burn yourself out. 

With any change to SymPy, and especially changes like this one, it's very important that we make sure that everything is mathematically correct. Mathematically wrong results are the worst sort of bug a computer algebra system can have, so we want to avoid them. This means that we are going to try to scrutinize PRs carefully. 

Personally I think it's fine if some function features are missing in the first pass, especially things like series expansion which often isn't easy to implement, considering how many methods SymPy has for it.

> Is it possible to define things like derivatives for harmonic? This rewrite indicates that it might be in some cases:

Be careful. Derivatives of Piecewise currently aren't correct for pointwise definitions. https://github.com/sympy/sympy/issues/11402. If you want to define a derivative that's a Piecewise you should try to find a closed-form expression for it.

oscarbenjamin commented:
> > Is it possible to define things like derivatives for harmonic? This rewrite indicates that it might be in some cases:
> 
> Be careful. Derivatives of Piecewise currently aren't correct for pointwise definitions.

The rule now implemented in this PR is:
```python
In [1]: harmonic(x, y).diff(x)
Out[1]: y⋅ζ(y + 1, x + 1)

In [2]: harmonic(x, y).diff(y)
Out[2]: 
∂                 
──(harmonic(x, y))
∂y    
```
So only differentiation wrt the first argument is defined (at least for arbitrary `y`). The rewrite gives:
```python
In [3]: harmonic(x, y).rewrite(zeta)
Out[3]: 
⎧polygamma(0, x + 1) + γ  for y = 1
⎨                                  
⎩  ζ(y) - ζ(y, x + 1)     otherwise

In [4]: harmonic(x, y).rewrite(zeta).diff(x)
Out[4]: 
⎧polygamma(1, x + 1)  for y = 1
⎨                              
⎩ y⋅ζ(y + 1, x + 1)   otherwise

In [5]: harmonic(x, y).rewrite(zeta).diff(y)
Out[5]: 
⎧            0               for y = 1
⎪                                     
⎨d          ∂                         
⎪──(ζ(y)) - ──(ζ(y, x + 1))  otherwise
⎩dy         ∂y                        
```
So at least for `y != 1` `Out[1]` is correct. When `y == 1` we have
```python
In [6]: harmonic(x, 1)
Out[6]: harmonic(x)

In [7]: harmonic(x, 1).rewrite(zeta)
Out[7]: polygamma(0, x + 1) + γ

In [8]: harmonic(x, 1).rewrite(zeta).diff(x)
Out[8]: polygamma(1, x + 1)

In [9]: harmonic(x, 1).rewrite(zeta).diff(x).rewrite(zeta)
Out[9]: ζ(2, x + 1)

In [10]: harmonic(x, y).diff(x).subs(y, 1)
Out[10]: ζ(2, x + 1)
```
So I think this seems to be all internally consistent. Note that we are differentiating wrt `x` but the Piecewise condition is for `y` so the problem in https://github.com/sympy/sympy/issues/11402 does not apply directly.

oscarbenjamin commented on sympy/functions/combinatorial/numbers.py:
```diff
@@ -866,12 +872,8 @@ class harmonic(Function):
-
-    which equals ``zeta(m)`` for ``m > 1``.
+    >>> limit(harmonic(n, m+1), n, oo) # does not hold for m == 1
+    zeta(m + 1)
```
Why doesn't this hold for `m == 1`?
```python
In [16]: limit(harmonic(n, m+1), n, oo)
Out[16]: ζ(m + 1)

In [17]: limit(harmonic(n, m+1), n, oo).subs(m, 1)
Out[17]: 
 2
π 
──
6 

In [18]: limit(harmonic(n, m+1).subs(m, 1), n, oo)
Out[18]: 
 2
π 
──
6 

In [19]: harmonic(n, m+1).subs(m, 1).subs(n, 1e6)
Out[19]: harmonic(1000000.0, 2)

In [20]: harmonic(n, m+1).subs(m, 1).subs(n, 1e6).n()
Out[20]: 1.64493306684873

In [21]: harmonic(n, m+1).subs(m, 1).subs(n, 1e12).n()
Out[21]: 1.64493406684723

In [22]: N(pi**2/6)
Out[22]: 1.64493406684823
```

Parcly-Taxel replied:
I was about to clarify in the documentation about this but you merged too soon. The comment `does not hold for m == 1` refers to $\lim_{n\to\infty}H_{n,m}=\zeta(m)$ not holding if the _second argument of `harmonic()` is 1, not the second argument of the given expression `harmonic(m+1)`.

oscarbenjamin replied:
We can update that in one of the other PRs then.

I see what you mean so the doctest uses `m+1` with `m` a positive integer because we want an integer greater than 1 for the second argument.

If the second argument is `1` then we get:
```python
In [9]: limit(harmonic(n, 1), n, oo)
Out[9]: ∞

In [10]: limit(harmonic(n, m+1), n, oo).subs(m, 0)
Out[10]: zoo
```

oscarbenjamin commented:
I had one question above (which is just a question and not a request for changes) but otherwise looking through I think this looks good.

oscarbenjamin commented:
Okay, let's get this one in.
